cmd/forwarder/auth.go

```diff
@@ -0,0 +1,212 @@
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+// credentials are used by basic auth and include the hash of a valid password, plus
+// a "stage" string which is used to emit metrics that are useful when managing credrolls, so that
+// we can track whether or not deprecated passwords are still in use.
+type credential struct {
+	Stage string `json:stage`
+	Hmac  string `json:hmac`
+}
+
+func newAuth(config AuthConfig, registry metrics.Registry) (*BasicAuth, error) {
+	if config.RedisUrl != "" && config.RedisKey == "" {
+		return nil, errors.New("RedisKey must be set if RedisUrl is set")
+	}
+
+	if config.RedisUrl == "" && config.Tokens == "" {
+		return nil, errors.New("At least one of RedisUrl or Tokens must be set.")
+	}
+
+	result, err := NewBasicAuthFromString(config.Tokens, config.HmacKey, registry)
+
+	if err != nil {
+		return result, err
+	}
+
+	if config.RedisUrl == "" {
+		return result, err
+	}
+
+	// Parse redis db out of url
+	opt, err := redis.ParseURL(config.RedisUrl)
+	if err != nil {
+		return result, err
+	}
+	client := redis.NewClient(opt)
+
+	// Refresh once at the start.
+	// Ignore errors here - if redis is down, we don't want the process to fail to start.
+	_, _ = result.refresh(client, config.HmacKey, config.RedisKey, config.Tokens)
```

Comment:
This can be dropped. The way `startRefresh` is written means auth.refresh will happen almost immediately.